### PR TITLE
Put sqs between upload api and adding to validation batch queue

### PR DESCRIPTION
**Pull request by parthshahva**
This PR puts sqs between upload api and validation batch queue for https://github.com/HumanCellAtlas/upload-service/issues/276

### CODE REVIEW

File: config/iam-policy-templates/upload-api-lambda.json [hunk 1]
```diff
@@ -79,7 +68,8 @@
       ],
       "Resource": [
         "arn:aws:sqs:us-east-1:$account_id:dcp-upload-pre-csum-queue-${DEPLOYMENT_STAGE}",
-        "arn:aws:sqs:us-east-1:$account_id:dcp-upload-area-deletion-queue-${DEPLOYMENT_STAGE}"
+        "arn:aws:sqs:us-east-1:$account_id:dcp-upload-area-deletion-queue-${DEPLOYMENT_STAGE}",
+        "arn:aws:sqs:us-east-1:$account_id:dcp-upload-pre-batch-validation-queue-${DEPLOYMENT_STAGE}"
```
Comment (sampierson) [inline config/iam-policy-templates/upload-api-lambda.json]: Why "pre" batch validation queue?  I guess this indicates that this is the queue before the validation process, and there may one day be a "post" batch validation queue, but I think names like "batch-validation-queue" and "batch-validation-results" would work better.

File: daemons/validation-scheduler-daemon/app.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    validation_id = unwrapped_event["validation_id"]
+    validator_image = unwrapped_event["validator_docker_image"]
+    environment = unwrapped_event["environment"]
+    orig_validation_id = unwrapped_event["orig_validation_id"]
+
+    upload_area = UploadArea(upload_area_uuid)
+    file = upload_area.uploaded_file(filename)
+    validation_scheduler = ValidationScheduler(file)
+    validation_id = validation_scheduler.schedule_batch_validation(validation_id,
+    															   validator_image,
```
Comment (sampierson) [inline daemons/validation-scheduler-daemon/app.py]: Seems like a copy-pasta error here.
This is a syntax error.

File: database/versions/c2edcbf1568d_add_scheduling_queued_to_validation.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+def upgrade():
+    op.execute('COMMIT')
+    op.execute("ALTER TYPE validation_event_status_enum ADD VALUE 'SCHEDULING_QUEUED';")
+    op.alter_column('validation', 'job_id', nullable=True)
+
+
+def downgrade():
+    op.execute('COMMIT')
+    op.execute("ALTER TYPE validation_event_status_enum DROP VALUE 'SCHEDULING_QUEUED';")
```
Comment (sampierson) [inline database/versions/c2edcbf1568d_add_scheduling_queued_to_validation.py]: Weird I was told removing items from enums was hard.  It is this easy?

File: terraform/modules/upload-service/secrets.tf [hunk 4]
```diff
@@ -18,7 +18,8 @@ resource "aws_secretsmanager_secret_version" "secrets" {
   "api_key": "${var.ingest_api_key}",
   "slack_webhook": "${var.slack_webhook}",
   "staging_bucket_arn": "${var.staging_bucket_arn}",
-  "ingest_api_host": "${var.ingest_api_host}"
+  "ingest_api_host": "${var.ingest_api_host}",
+  "pre_batch_validation_q_url": "${aws_sqs_queue.pre_batch_validation_queue.id}"
```
Comment (sampierson) [inline terraform/modules/upload-service/secrets.tf]: Same - nix the "pre"

File: terraform/modules/upload-service/sqs.tf [hunk 5]
```diff
@@ -76,3 +76,24 @@ resource "aws_lambda_event_source_mapping" "area_deletion_event_source_mapping"
   enabled           = true
   function_name     = "${aws_lambda_function.area_deletion_lambda.arn}"
 }
+
+resource "aws_sqs_queue" "pre_batch_validation_queue" {
```
Comment (sampierson) [inline terraform/modules/upload-service/sqs.tf]: -pre

File: tests/unit/__init__.py [hunk 6]
```diff
@@ -178,6 +179,7 @@ def setUp(self):
         self.sqs = boto3.resource('sqs')
         self.sqs.create_queue(QueueName=f"bogo_url")
         self.sqs.create_queue(QueueName=f"delete_sqs_url")
+        self.sqs.create_queue(QueueName=f"test_pre_batch_validation_q_url")
```
Comment (sampierson) [inline tests/unit/__init__.py]: -pre

File: tests/unit/common/test_validation_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        self.upload_area_id = str(uuid.uuid4())
+        self.upload_area = UploadArea(self.upload_area_id)
+        self.upload_area.update_or_create()
+
+    def tearDown(self):
+        super().tearDown()
+        pass
+
+    @patch('upload.common.upload_area.UploadedFile.size', MAX_FILE_SIZE_IN_BYTES + 1)
+    def test_check_file_can_be_validated_returns_false_if_file_is_too_large_for_validation(self):
```
Comment (sampierson) [inline tests/unit/common/test_validation_scheduler.py]: please start using my double-underscore technique:

`def test_<thing_under_test>__<scenario>__<should_do>`
e.g.
`def test_check_file_can_be_validated__when_file_is_too_large_for_validation__returns_false`

Comment (sampierson) [reply tests/unit/common/test_validation_scheduler.py]: I guess this is just a move of code, but it's a good time to revise it.

Comment (sampierson) [reply tests/unit/common/test_validation_scheduler.py]: This is not a technique that works everywhere, but it's good here.

File: daemons/validation-scheduler-daemon/Makefile [hunk 8]
```diff
@@ -0,0 +1,39 @@
+include ../../common.mk
+.PHONY: install build stage deploy clobber
+
+ZIP_FILE=validation_scheduler_daemon.zip
+BUCKET=$(BUCKET_NAME_PREFIX)lambda-deployment-$(DEPLOYMENT_STAGE)
+STAGED_FILE_KEY=$(ZIP_FILE)
+
+default: build
+
```
Comment (MDunitz) [inline daemons/validation-scheduler-daemon/Makefile]: Maybe not in this pr, but at some point we should combine the makefiles for all the lambdas. I think we could put everything other than deploy and the zip_file name in common.mk

File: config/iam-policy-templates/upload-api-lambda.json (hunk 1 above)
Comment (sampierson) [reply config/iam-policy-templates/upload-api-lambda.json]: Actually all validation is done in Batch, so nix that too.
Name it "dcp-upload-validation-queue"

File: daemons/validation-scheduler-daemon/Makefile (hunk 8 above)
Comment (parthshahva) [reply daemons/validation-scheduler-daemon/Makefile]: Great idea! Will create a ticket

File: config/iam-policy-templates/upload-api-lambda.json (hunk 1 above)
Comment (parthshahva) [reply config/iam-policy-templates/upload-api-lambda.json]: Sounds good

File: daemons/validation-scheduler-daemon/app.py (hunk 2 above)
Comment (parthshahva) [reply daemons/validation-scheduler-daemon/app.py]: Seemed to work when deployed but fixed to pass linting

File: database/versions/c2edcbf1568d_add_scheduling_queued_to_validation.py (hunk 3 above)
Comment (parthshahva) [reply database/versions/c2edcbf1568d_add_scheduling_queued_to_validation.py]: I think I was sleeping when I wrote the downgrade. It's not simple and not supported. Changing to the link to the stackoverflow to match similar migrations

File: tests/unit/common/test_validation_scheduler.py (hunk 7 above)
Comment (parthshahva) [reply tests/unit/common/test_validation_scheduler.py]: fixed all tests to follow this pattern in test_validation_scheduler.py

**Pull request merged**